File: site/app/data/Pokemon.js
class Pokemon {
  constructor(key, config, moves) {
    this.name = key;
    this.health = config.hp;
    this.attack = config.atk;
    this.defense = config.def;
    this.specialAttack = config.spAtk;
    this.specialDefense = config.spDef;
    this.speed = config.speed;
    this.quickMoves = config.quickMoves.map(name => moves[name]);
    this.chargeMoves = config.chargeMoves.map(name => moves[name]);
    this.types = config.types;
    this.atkIv = 15;
    this.defIv = 15;
    this.staIv = 15;
  }

  getAttack(iv = this.atkIv) {
    const higherAtk = Math.max(this.attack, this.specialAttack);
    const lowerAtk = Math.min(this.attack, this.specialAttack);
    const scaledAttack = Math.round(2 * (7 * higherAtk / 8 + lowerAtk / 8));
    return Math.round(scaledAttack * this.getSpeedMod() + iv);
  }

  getStamina(iv = this.staIv) {
    return (this.health * 1.75 + 50 + iv) | 0
  }

  getDefense(iv = this.defIv) {
    const higherDef = Math.max(this.defense, this.specialDefense);
    const lowerDef = Math.min(this.defense, this.specialDefense);
    const scaledDefense = Math.round(2 * (5 * higherDef / 8 + 3 * lowerDef / 8));
    return Math.round(scaledDefense * this.getSpeedMod() + iv);
  }

  getSpeedMod() {
    return 1 + (this.speed - 75) / 500;
  }
}

export default Pokemon

File: src/com/sucy/go/Pokemon.java
package com.sucy.go;

import java.util.List;

/**
 * PokemonGo © 2018
 * com.sucy.go.Pokemon
 */
public class Pokemon {
    public String     name;
    public int        health;
    public int        attack;
    public int        defense;
    public int        specialAttack;
    public int        specialDefense;
    public int        speed;
    public List<Move> quickMoves;
    public List<Move> chargeMoves;
    public List<Type> types;

    public int attack() {
        final int higherAtk = Math.max(attack, specialAttack);
        final int lowerAtk = Math.min(attack, specialAttack);
        final double scaledAttack = Math.round(2 * (7 * higherAtk / 8 + lowerAtk / 8));
        return (int) Math.round(scaledAttack * speedMod() + 15);
    }

    public int stamina() {
        return (int) Math.floor(health * 1.75 + 65);
    }

    public int defense() {
        final int higherDef = Math.max(defense, specialDefense);
        final int lowerDef = Math.min(defense, specialDefense);
        final double scaledDefense = Math.round(2 * (5 * higherDef / 8 + 3 * lowerDef / 8));
        return (int) Math.round(scaledDefense * speedMod() + 15);
    }

    private double speedMod() {
        return 1 + (speed - 75) / 500.0;
    }
}
